kotlinx-coroutines-core/common/test/flow/terminal/FirstTest.kt

```diff
@@ -173,4 +175,19 @@ class FirstTest : TestBase() {
 
         assertFailsWith<CancellationException> { flow.first() }
     }
-}
+
+    @Test
+    fun test() = runTest {
+        val job = launch(start = UNDISPATCHED) {
+            flow {
+                coroutineScope {
+                    emitAll(produce(start = UNDISPATCHED) {
+                        channel.send(Unit)
+                    })
+                }
+            }.first()
+            expectUnreached()
+        }
+        job.cancel()
+    }
```

Comment:
In this scenario, it's worth it to use `expect` and `finish` to assert the order of execution. The rules:
* `expect(1)` is the first invocation in any test.
* `expect(n + 1)` can only be called after `expect(n)`.
* `finish(n + 1)` can only be called after `expect(n)`.
* Each test that uses an `expect` must call `finish`.
* After `finish`, no extra `expect` or `finish` can be called.

With this, we can ensure that indeed, the test checks what we want it to check: first, cancellation happens, then, `first` throws a `CancellationException`.